presto-main/src/main/java/com/facebook/presto/memory/DefaultQueryContext.java

```diff
@@ -196,7 +196,11 @@ private synchronized boolean tryUpdateUserMemory(String allocationTag, long delt
     {
         if (delta <= 0) {
             ListenableFuture<?> future = updateUserMemory(allocationTag, delta);
-            verify(future.isDone(), "future should be done");
+            // When delta == 0 and the pool is full the future can still not be done,
```

Comment:
@nezihyigitbasi What does it mean that the future is not done (for delta = 0)? Is it safe for the caller to proceed? Would you explain why?

Reply:
> What does it mean that the future is not done (for delta = 0)?

This means the pool is full, if you try to allocate a positive delta then it will not be satisfied (return false if `tryUpdateUserMemory` is called, or fail if `updateUserMemory` is called).

> Is it safe for the caller to proceed? Would you explain why?

I think it is safe, because a delta of 0 means don't allocate or release anything (don't change the state of the pools/contexts), so when we return `true` here we just tell the caller that the call succeeded, which is really the case (see the contract in `LocalMemoryContext.trySetBytes) javadoc). If there are subsequent calls to `tryUpdateUserMemory` with delta = 0, they will behave similarly, so no issue with that too. If there is a subsequent call to `tryUpdateUserMemory` with a positive value then we will just catch it with the condition: `if (queryMemoryContext.getUserMemory() + delta > maxUserMemory)` and return false when the pool is full. If there is a subsequent call to other allocations method like `updateUserMemory` with a positive delta, then they will just fail as the pool is full (`EXCEEDED_LOCAL_MEMORY_LIMIT`).